src/polyfills/formdata-entries.js

```diff
@@ -0,0 +1,182 @@
+ * @this {FormData}
+ */
+function FormData(form) {
+  this.map_ = {};
+
+  if (!form) {
+    return new self.FormDataNative_(form);
+  }
+
+  for (let i = 0; i < form.elements[i]; i++) {
+    const control = form.elements[i];
+
+    if (!control.name) {
+      continue;
+    }
+
+    switch (control.type) {
+      case 'select-multiple':
+      case 'select-one':
+        Array.from(control.selectedOptions)
+            .forEach(option => this.append_(name, option.value));
+        break;
+      case 'checkbox':
+      case 'radio':
+        if (control.checked) {
+          this.append_(name, control.value);
+        }
+        break;
+      default:
+        this.append_(name, control.value);
+    }
+  }
+
+  return new self.FormDataNative_(form);
+}
+
+/**
+ * Appends a new value onto an existing key inside a `FormData` object, or adds
+ * the key if it does not already exist.
+ *
+ * This patches the native `append` method so that the polyfill can capture the
+ * field passed to the method. If `FormData#entries` needs a polyfill, chances
+ * are there are no methods to read the content of the `FormData` after
+ * construction, so the only way to implement `entries` polyfill is to patch
+ * this method (and the constructor).
+ *
+ * Since AMP doesn't support `<input type="file">`, appending a `File` object
+ * is not supported and the `filename` parameter is ignored for this polyfill.
+ *
+ * For more details on this, see http://mdn.io/FormData/append.
+ *
+ * @param {string} name The name of the field whose data is contained in
+ *     `value`.
+ * @param {string} value The field's value.
+ * @this {FormData}
+ */
+function append(name, value, filename) {
+  this.append_(name, value);
+  return this.appendNative_(name, value, filename);
+}
+
+/**
+ * Appends a new value onto an internal map used for the polyfill, or adds
+ * the key if it does not already exist.
+ *
+ * @param {string} name The name of the field whose data is contained in
+ *     `value`.
+ * @param {string} value The field's value.
+ * @this {FormData}
+ */
+function append_(name, value) {
+  this.map_[name] = this.map_[name] || [];
+  this.map_[name].push(value);
+}
+
+/**
+ * Returns an array of all key/value pairs contained in this object.
+ *
+ * By specification, `entries` returns an iterator instead of an array. However,
+ * returning an iterator requires writing an ES6 generator, which is not
+ * supported with the current Babel configuration (need the regenerator
+ * runtime). As a workaround, this polyfill returns an array, and callers are
+ * expected to wrap the return value with `Array.from(...)` so that an array can
+ * be obtained from both the iterator returned by the native `entries` function
+ * and the array returned by this polyfill.
+ *
+ * For more details on this, see http://mdn.io/FormData/entries.
+ *
+ * @return {!Array<!Array<string>>}
+ * @this {FormData}
+ */
+function entries() {
```

Comment:
Unless we add a presubmit check to enforce this, let's use a different function name. `entriesArray`?